Add ability to map unknown exceptions to custom exceptions

Brutus5000 opened the pull request:
Resolves #2204

## Description
ElideSettings now take an (optional) ErrorMapper. This is a functional interface that can map any Exception into a different CustomErrorException.

When handling a request now, Elide runs throw the previous error handling up to the point where it is considered an unknown exception. In that case, the unknown exception is passed to the error mapper, which can map it into a CustomErrorException. If the mapping is successful, we build an error response out of it, otherwise we rethrow the untouched exception (just as before).

<!--- Describe your changes in detail -->
<!--- If suggesting a new feature or change, please discuss it in an issue first -->
<!--- If fixing a bug, there should be an issue describing it with steps to reproduce -->

## Motivation and Context
While Elide maps javax.validation.ConstraintValidationException into proper error messages, it fails to do the same for org.hibernate.ConstraintViolationException as the latter ones might only refer to database constraint names, which have no context.
With a custom mapping a developer can capture these exceptions, read them out and turn them into a more meaningful exception (e.g. a tablename_unique_name constraint becomes a UsernameNotUniqueException extending CustomErrorException).

## How Has This Been Tested?
Not yet. This is only RFC so far.
<!--- Please describe in detail how you tested your changes. -->
<!--- Include details of your testing environment, and the tests you ran to -->
<!--- see how your change affects other areas of the code, etc. -->


## License
I confirm that this contribution is made under an Apache 2.0 license and that I have the authority necessary to make this contribution on behalf of its copyright owner.

aklish commented:
This looks good.  Will there be a default mapper and what will it do?

Brutus5000 commented:
No, by default the mapper is null, and if it's null no mapping takes place - in that case you have the same behavior as before this change.

aklish commented:
This looks good.  In the past, we have made these kinds of interfaces beans for Spring boot and elide standalone.   

Testing though is all that's needed for merging.

Brutus5000 commented:
FYI: There is currently no activity due to lack of time, but I for sure plan on adding tests and finish this PR in the next few months.

aklish commented:
> FYI: There is currently no activity due to lack of time, but I for sure plan on adding tests and finish this PR in the next few months.

Sounds good.  Thanks for the update.

Brutus5000 commented:
@aklish I needed to do some more refactorings in order to get it right. And I also added tests now. Do you think it is sufficient to just test one method, e.g. POST, and leave out the other http methods? (They all end up in the same handleRequest error handling anyway)

aklish commented:
@Brutus5000 One POST test should be fine.

Brutus5000 commented:
OK, so from my side it's done, but the CI checks seem to be frozen.

aklish commented:
> OK, so from my side it's done, but the CI checks seem to be frozen.

Reviewing now.  CI from non-writers requires a manually start in screwdriver to kickoff the build.

aklish commented:
Code changes look good - no comments on what was changed. 

To make this feature ubiquitous for Elide, two things need to be added:

1. Make the same changes made in Elide.java to QueryRunner.java for GraphQL.
2. Make similar changes made in ElideAutoConfiguration to ElideStandaloneSettings for ElideStandalone.

Brutus5000 commented:
@aklish I added the changes for StandaloneElide and QueryRunner.
For standalone I didn't add particular checks, but I added an "empty" error mapper to the test configuration.
For GraphQL QueryRunner all I can do it verify if the mapping is called. The test setup doesn't offer much more, as this part doesn't seem to be covered by unit tests.
